sync/rpc/server.go

```diff
@@ -282,27 +283,77 @@ func (*SyncAgentServer) BackupRemove(ctx context.Context, req *BackupRemoveReque
 	return &Empty{}, nil
 }
 
-func (*SyncAgentServer) BackupRestore(ctx context.Context, req *BackupRestoreRequest) (*Empty, error) {
-	cmd := reexec.Command("sbackup", "restore", req.Backup, "--to", req.SnapshotFileName)
-	cmd.SysProcAttr = &syscall.SysProcAttr{
-		Pdeathsig: syscall.SIGKILL,
+func (s *SyncAgentServer) anyActiveRestoreInProgress() bool {
+	s.Lock()
+	defer s.Unlock()
+
+	if s.RestoreList != nil && s.RestoreList.RestoreProgress != 100 && s.RestoreList.RestoreError == nil {
+		return true
 	}
-	cmd.Stdout = os.Stdout
-	cmd.Stderr = os.Stderr
-	if err := cmd.Start(); err != nil {
-		return nil, err
+	return false
+}
+
+func (s *SyncAgentServer) BackupRestore(ctx context.Context, req *BackupRestoreRequest) (*Empty, error) {
+	if s.anyActiveRestoreInProgress() {
+		return nil, fmt.Errorf("cannot initate backup restore as there is one already in progress")
 	}
 
-	logrus.Infof("Running %s %v", cmd.Path, cmd.Args)
-	if err := cmd.Wait(); err != nil {
-		logrus.Infof("Error running %s %v: %v", "sbackup", cmd.Args, err)
+	backupType, err := util.CheckBackupType(req.Backup)
+	if err != nil {
 		return nil, err
 	}
+	// set aws credential
+	if backupType == "s3" {
+		credential := req.Credential
+		// validate environment variable first, since CronJob has set credential to environment variable.
+		if credential != nil && credential[types.AWSAccessKey] != "" && credential[types.AWSSecretKey] != "" {
+			os.Setenv(types.AWSAccessKey, credential[types.AWSAccessKey])
+			os.Setenv(types.AWSSecretKey, credential[types.AWSSecretKey])
+			os.Setenv(types.AWSEndPoint, credential[types.AWSEndPoint])
+		} else if os.Getenv(types.AWSAccessKey) == "" || os.Getenv(types.AWSSecretKey) == "" {
+			return nil, errors.New("could not backup to s3 without setting credential secret")
+		}
+	}
 
-	logrus.Infof("Done running %s %v", "sbackup", cmd.Args)
+	rObj, err := backup.DoBackupRestore(req.Backup, req.SnapshotFileName)
+	if err != nil {
+		return nil, fmt.Errorf("error initiating restore [%v]", err)
+	}
+	logrus.Infof("Successfully initiated restore for snapshot [%v]", rObj.SnapshotName)
+
+	s.Lock()
```

Comment:
The check for existing and set have to be in the same critical area, otherwise there are things can happen between check and set.